akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldResourceAsyncSinkSpec.scala

```diff
@@ -0,0 +1,246 @@
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(_.contains("b")).mkString)
+    }
+
+    "continue when Strategy is Resume and exception happened asynchronously" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (r, s) ⇒ Future(if (s.contains("b")) throw TE("")).flatMap(_ ⇒ r.write(s)),
+        _.close()
+      )
+        .withAttributes(supervisionStrategy(resumingDecider))
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(_.contains("b")).mkString)
+    }
+
+    "continue when Strategy is Restart and exception happened synchronously" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (r, s) ⇒ if (s.contains("b")) throw TE("") else r.write(s),
+        _.close()
+      )
+        .withAttributes(supervisionStrategy(restartingDecider))
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(line ⇒ line.contains("a") || line.contains("b")).mkString)
+
+    }
+
+    "continue when Strategy is Restart and exception happened asynchronously" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (r, s) ⇒ Future(if (s.contains("b")) throw TE("")).flatMap(_ ⇒ r.write(s)),
+        _.close()
+      )
+        .withAttributes(supervisionStrategy(restartingDecider))
+      val fut = Source.fromIterator(() ⇒ manyLinesArray.iterator).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
+      (r.isOpened, r.isClosed) should be((true, true))
+      r.written.mkString should be(manyLinesArray.filterNot(line ⇒ line.contains("a") || line.contains("b")).mkString)
+    }
+
+    "fail when upstream fails" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](() ⇒ r.open(), _ write _, _.close)
+      val (probe, fut) = TestSource.probe[String].toMat(sink)(Keep.both).run()
+      probe.sendError(TE(""))
+      Await.result(fut.failed, remainingOrDefault) should be(TE(""))
+      (r.isOpened, r.isClosed) should be((true, true))
+    }
+
+  }
```

Comment:
I think there are scenarios not covered here, from the top of my head (but not exhaustive) happy path with an empty stream (`Source.empty`), running with a source that fails directly (`Source.failed`), failure in open, failure in close, `failure` in both open and `writeData`, failing/completing upstream before writeData completes or fails, closing upstream before open completes.